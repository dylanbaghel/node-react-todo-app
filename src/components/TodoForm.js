import React from 'react';
import classnames from 'classnames';

class TodoForm extends React.Component {
    constructor(props) {
        super(props);
        this.handleTextChange = this.handleTextChange.bind(this);
        this.onSubmit = this.onSubmit.bind(this);
        this.state = {
            text: '',
            errors: {}
        };
    }

    handleTextChange(e) {
        let text = e.target.value;
        this.setState(() => ({
            text
        }));
    }

    onSubmit(e) {
        e.preventDefault();
        if (!this.state.text.trim()) {
            this.setState(() => ({
                errors: {text: 'Please Enter Todo'}
            }));
            return;
        }
        this.setState(() => ({
            text: ''
        }));
        this.props.onFormSubmit({
            text: this.state.text
        });
    }

    render() {
        return(
            <div>
                <form onSubmit={this.onSubmit}>
                    <div className="form-group">
                        <input
                            type="text"
                            placeholder="Enter Todo....."
                            value={this.state.text}
                            onChange={this.handleTextChange}
                            className={classnames('form-control', {
                                'is-invalid': this.state.errors.text
                            })}
                        />
                        <div className="invalid-feedback">
                            {this.state.errors.text}
                        </div>
                    </div>
                    <button className="btn btn-primary">
                        Add
                    </button>
                </form>
            </div>
        );
    }
}

export default TodoForm;