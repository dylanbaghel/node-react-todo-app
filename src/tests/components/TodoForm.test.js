import React from 'react';
import { shallow } from 'enzyme';

import TodoForm from './../../components/TodoForm';

test('should render todo form correctly', () => {
    const wrapper = shallow(<TodoForm />);
    expect(wrapper).toMatchSnapshot();
});

test('should render error with no input', () => {
    const wrapper = shallow(<TodoForm />);
    expect(wrapper).toMatchSnapshot();
    wrapper.find('form').simulate('submit', {
        preventDefault: () => {}
    });
    expect(wrapper.state('errors').text).toBeDefined();
    expect(wrapper).toMatchSnapshot();
});

test('should set text state on change', () => {
    const value = 'My New Todo';
    const wrapper = shallow(<TodoForm />);
    wrapper.find('input').simulate('change', {
        target: { value }
    });
    expect(wrapper.state('text')).toBe(value);
});

test('should call onFormSubmit Prop on Valid Data', () => {
    const text = 'My New Todo'
    const onFormSubmitSpy = jest.fn();
    const wrapper = shallow(<TodoForm onFormSubmit={onFormSubmitSpy} />);
    wrapper.setState(() => ({
        text
    }));
    wrapper.find('form').simulate('submit', {
        preventDefault: () => {}
    });
    expect(wrapper.state()).toEqual({
        text: '',
        errors: {}
    });
    expect(onFormSubmitSpy).toHaveBeenLastCalledWith({
        text
    });
});