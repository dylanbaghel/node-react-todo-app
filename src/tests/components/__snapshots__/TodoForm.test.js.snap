// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render error with no input 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <div
      className="form-group"
    >
      <input
        className="form-control"
        onChange={[Function]}
        placeholder="Enter Todo....."
        type="text"
        value=""
      />
      <div
        className="invalid-feedback"
      />
    </div>
    <button
      className="btn btn-primary"
    >
      Add
    </button>
  </form>
</div>
`;

exports[`should render error with no input 2`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <div
      className="form-group"
    >
      <input
        className="form-control is-invalid"
        onChange={[Function]}
        placeholder="Enter Todo....."
        type="text"
        value=""
      />
      <div
        className="invalid-feedback"
      >
        Please Enter Todo
      </div>
    </div>
    <button
      className="btn btn-primary"
    >
      Add
    </button>
  </form>
</div>
`;

exports[`should render todo form correctly 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <div
      className="form-group"
    >
      <input
        className="form-control"
        onChange={[Function]}
        placeholder="Enter Todo....."
        type="text"
        value=""
      />
      <div
        className="invalid-feedback"
      />
    </div>
    <button
      className="btn btn-primary"
    >
      Add
    </button>
  </form>
</div>
`;
